### PR TITLE
legion: Add 24.03.0, update HIP dependency

### PR DESCRIPTION
**Pull request by elliottslaughter**
This PR adds:

 * Legion 24.03.0
 * HIP is now an explicit dependency
 * Put explicit bounds on the HIP version(s) supported
   * Note that prior to 24.03.0, Legion did not support ROCm 6.0, so there is no point in trying to patch it over
   * I can't remember the exact version where we added the first HIP support, but it was not always there, so there's no point in going to arbitrarily older versions either

I'd appreciate a check on the syntax as I am not a Spack expert.

### CODE REVIEW

File: var/spack/repos/builtin/packages/legion/package.py [hunk 1]
```diff
@@ -55,6 +56,9 @@ class Legion(CMakePackage, ROCmPackage):
     depends_on("cuda@10.0:11.9", when="+cuda @21.03.0:23.03.0")
     depends_on("cuda@10.0:12.2", when="+cuda_unsupported_compiler")
     depends_on("cuda@10.0:12.2", when="+cuda")
+    depends_on("hip@3.0:5.7", when="+rocm @23.03.0:23.12.0")
+    depends_on("hip@4.0:6.0", when="+rocm @24.03.0:")
+    depends_on("hip@4.0:6.0", when="+rocm @stable")
```
Comment (rbberger) [inline var/spack/repos/builtin/packages/legion/package.py]: Spack doesn't have any hip/rocm versions prior to 5.1.0 anymore.

**Comment by rbberger** [comment]
@elliottslaughter while you're at it, can you bump up the allowable CUDA versions to 12.4? Why is it limited in the first place? Let's say there was a problem with a newly released CUDA and Legion fails with it. Users would still be able to limit their cuda version during concretization. With a limit like there is at the moment, it forces anyone who is testing the bleeding edge make manual edits to this spackage. For my CI deployments I have to replace it completely, since inheritance isn't good enough to remove a restriction like this.

File: var/spack/repos/builtin/packages/legion/package.py [hunk 2]
```diff
@@ -350,11 +353,10 @@ def cmake_args(self):
...
             options.append(from_variant("Legion_HIP_ARCH", "amdgpu_target"))
             options.append(from_variant("Legion_HIJACK_HIP", "hip_hijack"))
-            options.append(self.define("HIP_PATH", "{0}/hip".format(spec["hip"].prefix)))
-            if "^hip@:5.7" in spec:
+            if "@23.03.0:23.12.0" in spec:
                 options.append(self.define("HIP_PATH", "{0}/hip".format(spec["hip"].prefix)))
-            elif "^hip@6.0:" in spec:
-                options.append(self.define("HIP_PATH", "{0}".format(spec["hip"].prefix)))
+            else:
+                options.append(self.define("ROCM_PATH", "{0}".format(spec["hip"].prefix)))
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/legion/package.py]: FYI. As we update the core, we are also encouraging use of Python f-strings in packages (see https://docs.python.org/3/tutorial/inputoutput.html).

**Review by tldahlgren — COMMENTED** [review]
Confirmed the new version's commit exists.

File: var/spack/repos/builtin/packages/legion/package.py (hunk 2 above)
Comment (rbberger) [reply var/spack/repos/builtin/packages/legion/package.py]: In this case, you can even avoid the format altogether
```suggestion
                options.append(self.define("ROCM_PATH",  spec["hip"].prefix))
```

**Comment by elliottslaughter** [comment]
@rbberger @tldahlgren Please look at the latest version and see what you think.

File: var/spack/repos/builtin/packages/legion/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/legion/package.py]: And everything prior to 5.3.0 is deprecated (see 'spack info hip').

**Review by tldahlgren — DISMISSED** [review]
Much cleaner, thanks!

**Comment by tldahlgren** [comment]
@spackbot fix style

**Review by rbberger — APPROVED** [review]
Nice. Thank you. LGTM.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**